### PR TITLE
Load ClientCertificateMode from config

### LINKED ISSUE #18660: Setting the ClientCertificateMode Kestrel server option from an appsettings.json file.
Creating this issue was recommended to me from a response to the [AspNetCore.Docs issue #16759](https://github.com/aspnet/AspNetCore.Docs/issues/16759) I created previously.

**Context:** From the Microsoft Docs entry for the [Kestrel web server implementation in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1), under the [Kestrel options](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#kestrel-options) section, the following is a description about configuring Kestrel options:

>Kestrel options, which are configured in C# code in the following examples, can also be set using a configuration provider. For example, the File Configuration Provider can load Kestrel configuration from an appsettings.json or appsettings.{Environment}.json file:

I would like to use the Kestrel web server for local development and IIS for the remote web hosting of an ASP.NET Core web application.

I have successfully followed the [Configure your host to require certificates](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1#configure-your-host-to-require-certificates) in `Program.cs` for the [Kestrel](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1#kestrel) web server.

**Issue:** It would be nice to be able to configure the `ClientCertificateMode` property of the `HttpsConnectionAdapterOptions` class from an _appsettings.json_ file, somewhat similar to how the `EndpointsDefault` property is configured in the _appsettings.json_ file, documented in the [Endpoint configuration](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#endpoint-configuration) section of the Kestrel Microsoft Docs entry. The `EndpointsDefault` property is a sibling property of the `HttpsDefaults` property in the `KestrelServerOptions` class, and I made an (incorrect) assumption that it would be configurable from an _appsettings.json_ file.

In the response to the previous Microsoft Docs issue I created, the [`ConfigurationReader.cs`](https://github.com/dotnet/aspnetcore/blob/master/src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs) class was linked, providing a _**Rosetta Stone**_ like tool for comprehending the mapping between the `KestrelServerOptions` [class members](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.core.kestrelserveroptions?view=aspnetcore-3.1) and the corresponding _appsettings.json_ configuration file keys and values. While the `HttpsDefaults` key is not found in the `ConfigurationReader` class like the `EndpointDefaults` key, might there be in a future update the addition of the `HttpsDefaults` key with the option to set the `ClientCertificateMode` configuration value?

cc: @guardrex

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/dotnet/aspnetcore/issues/18660

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs [hunk 1]
```diff
@@ -293,6 +293,8 @@ public void Load()
                     httpsOptions.ServerCertificate = LoadCertificate(endpoint.Certificate, endpoint.Name)
                         ?? httpsOptions.ServerCertificate;
 
+                    httpsOptions.ClientCertificateMode = ConfigurationReader.ClientCertificateMode ?? httpsOptions.ClientCertificateMode;
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:296]: Organization: This was inserted between to related code sections dealing with server certs. Move it up above line 293.

File: src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs [hunk 2]
```diff
@@ -100,6 +104,16 @@ private IEnumerable<EndpointConfig> ReadEndpoints()
             return endpoints;
         }
 
+        private ClientCertificateMode? ReadClientCertificateMode()
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:107]: This only supports a global setting, not a setting per endpoint?

Comment (Kahbazi) [reply src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:107]: I couldn't find the code which options for each endpoint is being set from config. Could you point me to it please?

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:107]: https://github.com/dotnet/aspnetcore/blob/050632502e4813e4e70344f7d469d602cb12d2a7/src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs#L98

Comment (Kahbazi) [reply src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:107]: Uhhmm. I'm confused! `EndpointConfig` doesn't have a `ClientCertificateMode` property. Am I missing something? Is it even possible to set `ClientCertificateMode` per endpoint?

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/ConfigurationReader.cs:107]: EndpointConfig is an internal config construct, you should add ClientCertificateMode there.

File: src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs [hunk 3]
```diff
@@ -289,9 +289,11 @@ public void Load()
                         httpsOptions.SslProtocols = endpoint.SslProtocols.Value;
                     }
 
+                    httpsOptions.ClientCertificateMode = endpoint.ClientCertificateMode ?? ConfigurationReader.ClientCertificateMode ?? httpsOptions.ClientCertificateMode;
+
                     // Specified
                     httpsOptions.ServerCertificate = LoadCertificate(endpoint.Certificate, endpoint.Name)
-                        ?? httpsOptions.ServerCertificate;
+                        ?? httpsOptions.ServerCertificate;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:296]: revert.

```suggestion
                        ?? httpsOptions.ServerCertificate;
```

File: src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs [hunk 4]
```diff
@@ -289,9 +289,11 @@ public void Load()
                         httpsOptions.SslProtocols = endpoint.SslProtocols.Value;
                     }
 
+                    httpsOptions.ClientCertificateMode = endpoint.ClientCertificateMode ?? ConfigurationReader.ClientCertificateMode ?? httpsOptions.ClientCertificateMode;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:292]: I think we should apply the global ConfigurationReader.ClientCertificateMode before calling ApplyHttpsDefaults then apply endpoint.ClientCertificateMode afterwards. This way global config doesn't override what was done in code via ConfigureHttpsDefaults. Our handling of SslProtocols is very similar.

We should also add some KestrelConfigurationLoader tests to verify whatever behavior we decide on.

**Review by halter73 — APPROVED** [review]

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks

**Comment by kamranayub** [comment]
You're the best @Kahbazi thank you, a small quality of life improvement but much appreciated 🥇

**Pull request merged**